[Babel 8] Refactor register cache

liuxingbaoyu opened the pull request:
<!--
Before making a PR, please read our contributing guidelines
https://github.com/babel/babel/blob/main/CONTRIBUTING.md

Please note that the Babel Team requires two approvals before merging most PRs.

For issue references: Add a comma-separated list of a [closing word](https://help.github.com/articles/closing-issues-via-commit-messages/) followed by the ticket number fixed by the PR. (it should be underlined in the preview if done correctly)

If you are making a change that should have a docs update: submit another PR to https://github.com/babel/website
-->

| Q                        | A <!--(Can use an emoji 👍) -->
| ------------------------ | ---
| Fixed Issues?            | Fixes #5667, Fixes #9320 <!-- remove the (`) quotes and write "Fixes" before the number to link the issues -->
| Patch: Bug Fix?          | √
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | √
| Documentation PR Link    | <!-- If only readme change, add `[skip ci]` to your commits -->
| Any Dependency Changes?  |
| License                  | MIT

<!-- Describe your changes below in as much detail as possible -->
This also fixes a hidden bug in Babel 8: `process.on("exit", cb)` would not be called.

nicolo-ribaudo commented on packages/babel-register/src/worker/cache.mts:
```diff
@@ -0,0 +1,102 @@
+      });
+
+      let createdAt = this.db.get("created_at");
+      // If the cache is older than 30 days, clear it
```
If I compile something every day for 31st day, on the 31st day will it delate all the cache or the data from the 1t day?

JLHwung commented on packages/babel-register/package.json:
```diff
@@ -17,7 +17,8 @@
     "clone-deep": "^4.0.1",
-    "find-cache-dir": "condition:BABEL_8_BREAKING ? ^3.3.1 : ^2.0.0",
+    "find-cache-dir": "condition:BABEL_8_BREAKING ? ^5.0.0 : ^2.0.0 (esm:default)",
+    "lmdb": "condition:BABEL_8_BREAKING ? ^3.2.6 : (esm:open)",
```
Can we make LMDB an optional dependency? The node-lmdb depends on node-gyp-build, which will kick it when the prebuilt binary bindings does not match the user arch. It is likely to happen when we still support a node version that is not supported by the prebuilt.

This opens all sort of install issues because of the tools/libs required and users may not be familiar with them.

See also https://github.com/parcel-bundler/parcel/issues/8152 for an example.

If we make it an optional dependency, users can still use Babel register regardless of node-gyp-errors, though the cache handling is not as efficient as it would be.

liuxingbaoyu replied:
I agree that this is an annoying problem.
But I worry that making it an optional dependency will make almost no one use it.

liuxingbaoyu replied:
I investigated and the pre-built packages released by lmdb support node-api, so the version shouldn't be a problem.
But they do not provide some pre-built packages compared to rollup, such as the lack of Windows x86 and arm.

liuxingbaoyu commented:
I changed it to `cacache`.
Since it uses a lot of small files, the performance may not be comparable to `lmdb`.
I added a batch cache logic, hoping the performance will be better.
It would be nice if `lmdb` could support more platforms, and `jest` would also benefit.

JLHwung commented:
> I changed it to `cacache`.
> 
> Since it uses a lot of small files, the performance may not be comparable to `lmdb`.
> 
> I added a batch cache logic, hoping the performance will be better.
> 
> It would be nice if `lmdb` could support more platforms, and `jest` would also benefit.

Thought: if the batch cacache is performant, we can use it for the Babel-loader as well.

liuxingbaoyu commented:
@JLHwung 
What do you think about importing webpack internal files directly or do we explicitly depend on a specific version of webpack to use its cache.
https://github.com/webpack/webpack/blob/eda2e284ac7020fb8e8aa5a6c1fd1547a0ad1b08/lib/cache/PackFileCacheStrategy.js
This is incredibly awesome and seems easy to use.

JLHwung commented:
@liuxingbaoyu The webpack cache PR was held back because of https://github.com/babel/babel/issues/16915#issuecomment-2423346921, but we don't have an opportunity to debug why the webpack cache is slower than the current babel-loader file cache.

nicolo-ribaudo commented on packages/babel-register/src/worker/babel-core.cts:
```diff
@@ -1,16 +1,16 @@
 
-exports.init = import("@babel/core").then(initialize);
+exports.init = (async function () {
+  await import("@babel/core").then(initialize).catch(err => {
```
Can we `require(esm)` now?
